java/src/main/java/com/microsoft/onnxruntime/ONNXAllocator.java

```diff
@@ -0,0 +1,294 @@
+/*
+ * Copyright © 2019, Oracle and/or its affiliates. All rights reserved.
+ */
+package com.microsoft.onnxruntime;
+
+import java.io.IOException;
+import java.nio.Buffer;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.DoubleBuffer;
+import java.nio.FloatBuffer;
+import java.nio.IntBuffer;
+import java.nio.LongBuffer;
+import java.nio.ShortBuffer;
+
+/**
+ * A wrapper around the onnx memory allocator.
+ */
+public class ONNXAllocator implements AutoCloseable {
+
+    final long handle;
+
+    private boolean closed = false;
+
+    static {
+        try {
+            ONNX.init();
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to load ONNX library",e);
+        }
+    }
+
+    /**
+     * Constructs a CPU allocator.
+     * @throws ONNXException If the onnx runtime threw an error.
+     */
+    public ONNXAllocator() throws ONNXException {
+        handle = createAllocator(ONNX.ortApiHandle);
+    }
+
+    ONNXAllocator(long handle) {
+        this.handle = handle;
+    }
+
+    /**
+     * Create a Tensor from a Java primitive (or String) multidimensional array.
+     * The shape is inferred from the array using reflection.
+     * @param data The data to store in a tensor.
+     * @return An ONNXTensor storing the data.
+     * @throws ONNXException If the onnx runtime threw an error.
+     */
+    public ONNXTensor createTensor(Object data) throws ONNXException {
```

Comment:
>Object data [](start = 35, length = 11)

Suggest a separate overload for String[] and non-string data

Reply:
So createTensor(String[]), createTensor(String) and createTensor(Object) or just the String[] and Object ones?

Reply:
String[] and something for raw data. Perhaps something better than an Object?

---
In reply to: [340868484](https://github.com/microsoft/onnxruntime/pull/2215#discussion_r340868484) [](ancestors = 340868484)

Reply:
There's no top type for arrays nor primitives, so it's hard to have a more specific type. There's also no top type for multidimensional primitive arrays that keep the primitive type but abstracts over the dimensionality. Until Java gets a good tensor type or reified generics I'm not sure what's a nicer type.

Reply:
So what does Object represent given all the overloads below?

---
In reply to: [340880169](https://github.com/microsoft/onnxruntime/pull/2215#discussion_r340880169) [](ancestors = 340880169)

Reply:
Object is anything that isn't matched by the overloads. Then the `TensorInfo.constructFromJavaArray` call throws an `ONNXException` if it's not a convertible type, if it is then it constructs an `ONNXTensor`.